content: draft: define source-track objective in terms of revisions and provenance. (#1083)
slsa-framework/slsa: docs/spec/draft/source-requirements.md

## Objective

The SLSA source track describes increasing levels of trustworthiness and completeness in a repository revision's provenance (e.g. how it was generated, who the contributors were, etc...).

The Source track is scoped to revisions of a single repository that is controlled by an organization.
That organization determines the intent of the software in the repository, what Source level should apply to the repository and administers technical controls to enforce that level.

The primary purpose of the Source track is to enable verification that the creation of a revision followed the expected process.
Consumers can examine the various source provenance attestations to determine if all sources used during the build meet their requirements

> Doing so prevents several classes of supply chain threats.

Reviewer: We should probably enumerate those at some point.  Is this the time to do so?

Author: yeah, this is the wording from the levels doc -- I think it's saying "these levels represent bundles of Good Ideas and are widely regarded by experts as effective at preventing classes of supply chain threats. Each increasing level is constructed to prevent specific threat vectors of increasing complexity."

Author: ```suggestion
```

## Objective

> The SLSA source track describes increasing levels of trustworthiness and completeness in a repository revision's provenance.

Reviewer: ```suggestion
The SLSA source track describes increasing levels of trustworthiness and completeness in a repository revision's provenance (e.g. how it was generated, who the contributors were, etc...).
```

> Source provenance describes which Source Control Platform (SCP) produced the revision, when it was generated, what process was used, and who were the contributors.

Reviewer: This seems out of place here.  When I read it it sounds like it's describing the attestation format but we haven't even finished talking about the objective yet.

I think the reason it's here is to explain 'provenance' for source in the sentence above.  I added a suggestion to just make that a parenthetical to get the point across without sounding like we're describing a specific format yet.

> Higher levels provide increasing protection against tampering with any parts of the Version Control System, Source Control Platform, or the source provenance.

Reviewer: ```suggestion
Higher source levels provide increasing protection against tampering with any parts of the Version Control System, Source Control Platform, or the source provenance.
```

Reviewer: Could also sub 'attestations' for 'information'.

> The primary purpose of the source track is to enable verification that the creation of a revision followed the expected process.
> Consumers have some way of knowing what the expected provenance should look like for a given revision and compare a revision's actual provenance to those expectations.

Reviewer: ```suggestion
The primary purpose of the source track is to enable verification that the process to create a revision meets the expected security properties.
Consumers have some way of mapping the sources listed in the _build_ provenance to the various _source_ provenance attestations to determine if all sources used during the build meet their requirements.
```

Author: the build provenance hasn't been computed yet to be used for this right? When you check source provenance, you're either looking at the build instructions themselves or trying to build some other artifact. 

At least in some cases, I would not expect a build to continue if one of the inputs fails to meet its requirements. This seems true whether the dependency is a package or repository.

Reviewer: Oh so I think this gets back to a conversation we were having someplace else.

Sometimes folks will want to examine the source attestations (VSAs?) _after_ the build with the build provenance (that's what we do).  Often the builder doesn't know what the policy for any given package should be so it's helpful to be able to defer until later.

However, I think the important point your making is that _sometimes_ (perhaps often) the builder will be checking these things and it doesn't need to compare it with the build provenance.

In that case we could either drop "Consumers have some way of mapping the sources listed in the _build_ provenance to the various _source_ provenance attestations to determine if all sources used during the build meet their requirements." or we could adjust it to "Consumers can examine the various _source_ provenance attestations to determine if all sources used during the build meet their requirements." (which is generic enough).

We'll probably need to come up with some model and options for "how & when to verify source provenance attestations" but that can happen later.

> The primary purpose of the source track is to enable verification that the creation of a revision followed the expected process.
> Consumers have some way of knowing what the expected provenance should look like for a given revision and compare a revision's actual provenance to those expectations.
> Doing so prevents several classes of supply chain threats.

Reviewer: ```suggestion
The primary purpose of the source track is to enable verification that the process to create a revision meets the expected security properties. Consumers have some way of mapping the sources listed in the _build_ provenance to the various _source_ provenance attestations to determine if all sources used during the build meet their requirements.
```

The Source track is scoped to revisions of a single repository that is controlled by an organization.
That organization determines the intent of the software in the repository, what Source level should apply to the repository and administers technical controls to enforce that level.

## Definitions

| Term | Description
| --- | ---
| Source | An identifiable set of text and binary files and associated metadata. Source is regularly used as input to a build system (see [SLSA Build Track](requirements.md)).
| Version Control System (VCS)| Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.
| Revision | A specific state of the source with an identifier provided by the version control system. As an example, you can identify a git revision by its tree hash.
| Source Control Platform (SCP) | A service or suite of services for hosting version-controlled software. GitHub and GitLab are examples of source control platforms, as are combinations of tools like Gerrit code reviews with GitHub source control.

> | Source Provenance | Verifiable information about a source revision describing where, when and how it was produced.

Reviewer: ```suggestion
| Source Provenance | Information  about which Source Control Platform (SCP) produced a revision, when it was generated, what process was used, who were the contributors, and what parent revisions it was based on.
```
I'm having second thoughts about cutting this.   In the objective section we seem to be using 'provenance' in the more generic sense (what is this code and what do we know about it).  Would it make sense to describe that concept here separate from whatever attestations carry that information?

It seems helpful to have the generic concept of provenance decoupled from the specific implementation (which would need to be verifiable).

Of course that might be at odds with [the 'provenance' definition here](https://slsa.dev/spec/v1.0/terminology#:~:text=by%20the%20build.-,Provenance,-Attestation%20(metadata)%20describing).

Seems like there should be some way to work it out.

| Organization | A collection of people who collectively create the Source. Examples of organizations include open-source projects, a company, or a team within a company. The organization defines the goals and methods of the repository.
| Repository | A uniquely identifiable instance of a VCS hosted on an SCP. The repository controls access to the Source in the version control system. The objective of a repository is to reflect the intent of the organization that controls it.

> | Change | A set of modifications to the source, or the computed difference between two revisions. A change is proposed by a [publisher](#source-roles), and applied to a specific revision to create a new revision. Change metadata MUST include any information required to situate the change in relation to other changes (e.g. parent revision).

Author: https://github.com/slsa-framework/slsa/pull/1083#discussion_r1669216998

update to talk about ref context

> | Change History | A record of the history of revisions that preceded a specific revision.

Author: I updated the text here a bit to help it cover git repos too (which don't track change history, technically).

> | Branch | A named pointer to a revision. Branches may be modified by authorized actors.

Author: added a section about branches / refs. needs to move above the "change" definition that will need to use it.

## Objective

The SLSA source track describes increasing levels of trustworthiness and completeness in a repository revision's provenance (e.g. how it was generated, who the contributors were, etc...).

The Source track is scoped to revisions of a single repository that is controlled by an organization.
That organization determines the intent of the software in the repository, what Source level should apply to the repository and administers technical controls to enforce that level.

The primary purpose of the Source track is to enable verification that the creation of a revision followed the expected process.

> Consumers can examine the various source provenance attestations to determine if all sources used during the build meet their requirements

Reviewer 2: ```suggestion
Consumers can examine the various source provenance attestations to determine if all sources used during the build meet their requirements.
```

## Definitions

| Term | Description
| --- | ---
| Source | An identifiable set of text and binary files and associated metadata. Source is regularly used as input to a build system (see [SLSA Build Track](requirements.md)).

> | Version Control System (VCS)| Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.

Reviewer 2: Should git be capitalised here? it's not capitalised in the next entry

Author: In this case, I think it's just because it's the beginning of the sentence.

| Revision | A specific state of the source with an identifier provided by the version control system. As an example, you can identify a git revision by its tree hash.

> | Source Control Platform (SCP) | A service or suite of services for hosting version-controlled software. GitHub and GitLab are examples of source control platforms, as are combinations of tools like Gerrit code reviews with GitHub source control.

Reviewer 2: ```suggestion
| Source Control Platform (SCP) | A service or suite of services (self-hosted or SaaS) for hosting version-controlled software. GitHub and GitLab are examples of source control platforms, as are combinations of tools like Gerrit code reviews with GitHub source control.
```
"Can we self host" is a frequent question, is it worth including in the definition?

> | Source Provenance | Information  about which Source Control Platform (SCP) produced a revision, when it was generated, what process was used, who the contributors were, and what parent revisions it was based on.

Reviewer 2: ```suggestion
| Source Provenance | Information about which Source Control Platform (SCP) produced a revision, when it was generated, what process was used, who the contributors were, and what parent revisions it was based on.
```

| Organization | A collection of people who collectively create the Source. Examples of organizations include open-source projects, a company, or a team within a company. The organization defines the goals and methods of the repository.
| Repository | A uniquely identifiable instance of a VCS hosted on an SCP. The repository controls access to the Source in the version control system. The objective of a repository is to reflect the intent of the organization that controls it.
| Branch | A named pointer to a revision. Branches may be modified by authorized actors. In git, cloning a repo will download all revisions in the history of the "default" branch to the local machine.
| Change | A set of modifications to the source in a specific context. As an example, a proposed change to a "releases/1" branch may require higher scrutiny than a change to "users/1".
| Change History | A record of the history of revisions that preceded a specific revision.

## Source Roles

| Role | Description
| --- | ---

> | Administrator | A human who can perform privileged operations on one or more projects. Privileged actions include, but are not limited to, modifying the change history and modifying project- or organization-wide security policies.

Reviewer 2: Nit: projects is not defined.

Author: Agree -- this PR only updates the objective (and required definitions). 
I think we're informally using "repository" instead of "project" for this concept, but we should discuss that in the next PR. 

To me, one of the reasons to use "project" is to encompass things like issues, actions, other stuff besides versioning. 
We're not finished, but I suspect this specification will need to focus entirely on source topics, ignoring everything else related to software development.

Reviewer 2: I agree with your suspicion, I think the other aspects of projects are out of scope for this track. I don't mind handling this in a future PR but we need a mechanism to track that the work still needs doing. I filed https://github.com/slsa-framework/slsa/issues/1099

| Version Control System (VCS)| Software for tracking and managing changes to source. Git and Subversion are examples of version control systems.
| Revision | A specific state of the source with an identifier provided by the version control system. As an example, you can identify a git revision by its tree hash.
| Source Control Platform (SCP) | A service or suite of services for hosting version-controlled software. GitHub and GitLab are examples of source control platforms, as are combinations of tools like Gerrit code reviews with GitHub source control.

> | Source Provenance | Information  about which Source Control Platform (SCP) produced a revision, when it was generated, what process was used, who were the contributors, and what parent revisions it was based on.

Reviewer: ```suggestion
| Source Provenance | Information  about which Source Control Platform (SCP) produced a revision, when it was generated, what process was used, who the contributors were, and what parent revisions it was based on.
```

| Trusted robot | Automation with an authentic identity that is authorized by the organization to propose and/or approve changes to the source.

> | Untrusted person | A human who has limited access to the project. They MAY be able to read the source. They MAY be able to propose or review changes to the source. They MAY NOT approve changes to the source or perform any privileged actions on the project.

Author: we might want to change all instances of "changes to the source" to "new revisions."

Reviewer: I see the benefit of changing some of those instances, but we probably want to keep some?

After all, quite a bit of the point is that actors _want_ to make "changes to the source" and one of the reasons we need the source track is to make that safe.  There can perhaps be a section that explains "Changes to the source are realized as new revisions within a repository" or something like that?

Author: Let's track it and skip for now.

Author: https://github.com/slsa-framework/slsa/issues/1095

| Proposer | The role that proposes a particular change to the source.

> | Reviewer | The role that reviews a particular change to the source.

Reviewer: ```suggestion
| Reviewer | A role that reviews a particular change to the source.
```

There can be multiple reviewers so I suggest 'a' instead of 'the'.
